### DISCUSSION TITLE
External ADC as source for CT power energy measurement

### DISCUSSION THREAD

**Discussion by hendrikvh** [arendst/Tasmota#21764] [Q&A] [answered]
Hi everyone, long time reader first poster.

I want to measure energy using a current transformer and external A/D converter instead of the internal ADC. Is this supported in Tasmota?

More info:
I'm using a ADS1115 as an external A/D and configured it via I2C. This works great and the values display as `ADS1115 Analog0` etc in the UI. I also know how to use the internal ADC as `CT power` to measure energy. However, how do I tell Tasmota to use the value from the external ADC as input to the CT power energy measurement?

I basically want to do what they describe here for ESPHome:

> You need to set up a voltage sensor source... for example ADS1115 and pass it to the CT clamp sensor with the sensor option.
https://esphome.io/components/sensor/ct_clamp.html

Thanks!

**Comment by sfromis** [comment, marked as answer]
The Tasmota feature for analog CT energy monitoring is only for the on-chip ADC of ESP8266 and ESP32.

**Reply by ortegafernando** [reply]
Hi, is there any way by berry script? As @hendrikvh has written code... will it be difficult to implement it in tasmota firmware? I have read a lot about "bad analog reading" from ESP8266 and ESP32. So a ADS1115 for low AC currents is a must, isn't it? 
Could @hendrikvh make a PR for this?

**Comment by hendrikvh** [comment]
For anyone else coming across this in the future: I added the ability [here](https://github.com/hendrikvh/Tasmota/compare/master...currentclamp) to

- Use and [ADS1115](https://www.ti.com/lit/ds/symlink/ads1113.pdf?ts=1721473434388) external ADC with a current clamp
- Use [RMS voltage equivalent ](https://www.electronics-tutorials.ws/accircuits/rms-voltage.html)for accurate power measurement rather than naive peak-to-peak estimation currently in use.
- Collect and average sample values over teleperiod when reporting

Willing to clean this up if anyone else ends up needing it. The on board ADC on ESP8266 and ESP32 was not linear enough for accurate values during my testing. Calibrated and confirmed accurate to the nearest ~20w against a known accurate utility meter in the range 100w - 10kW @ 230V.
